Ensure that deleter is called even for a no-data tensor.

aorenste opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117418

Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

aorenste committed:
Ensure that deleter is called even for a no-data tensor.

Summary:
The deleter is called indirectly by a std::unique_ptr<> whch only fires if there's actually data to delete.

Fixes #117273

Test Plan:

[ghstack-poisoned]

wjakob commented on aten/src/ATen/templates/Functions.cpp:
```diff
@@ -7,41 +7,47 @@
+  if (data_ == nullptr) {
+    // We need to ensure that there's always a valid pointer or the underlying
+    // std::unique_ptr<> won't call the deleter (custom or context).
+    data_ = malloc(0);
```
Won't this leak if not matched by a corresponding `free()` call? How about setting this to some dummy value like `(void *) 0x1` instead?

aorenste replied:
It's not a dummy value - it gets turned into a std::unique_ptr -- but tracing through the code again I'm not sure why (in the deleter case) it isn't actually freed twice since it's passed to two different std::unique_ptrs.  I'm going to convert this to a draft until I convince myself what is actually going on.

aorenste replied:
Ok - I see my misunderstanding. `UniqueVoidPtr(p)` doesn't actually delete `p` when it goes out of scope. You're right - this value is essentially a dummy which won't be deleted.

aorenste replied:
Updated to not use the dummy allocation - this is a cleaner solution.

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

wjakob commented:
LGTM, thanks!

aorenste commented:
@torchbot merge

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste committed:
Update on "Ensure that deleter is called even for a no-data tensor."


Summary:

When using a custom deleter InefficientStdFunctionContext was using a
std::unique_ptr<> to store the pointer and call the deleter - but this failed to
call the deleter if the pointer was null. Since we have a separate holder class
anyway take out the std::unique_ptr<> and call the deleter directly.

Fixes #117273

Test Plan:

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
